RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,358 @@
+`__torch_function__`, however, it only operates on functions and not on
+methods, and support for subclassed `torch.Tensor` objects in this protocol is
+limited.
+
+### Proposal
+We propose the following signature change to `__torch_function__`, to make it
+match NumPy: [[4]]
+
+```python
+class SubTensor(torch.Tensor):
+    def __torch_function__(self, func, types, args, kwargs):
+        # Implementation here
+```
+
+The reason for adding `types` to the signature is necessitated by the need for
+`super()`. If we set a requirement for `super().__torch_function__` to work
+properly, we would need to provide an easy way for users to signal to
+`__torch_function__` that they are calling to the next-specific implementation.
+
+### Process followed during a function/method call
+
+The process followed during a function/method call would be equivalent to:
+
+1. The dispatcher is called to extract the `Tensor`-likes.
+2. All `Tensor`-likes are checked for `__torch_function__`. If none exist, the
+   internal implementation is called, and the final result is returned.
+3. A collection of types that implement `__torch_function__` is created, with
+   no guaranteed order other than that subclasses come before superclasses.
+4. For one instance of each type in `types`, `__torch_function__` is called.
+   The first such function or method to return something other than
+   `NotImplemented` will be the final result. All exceptions will be propagated
+   upward.
+5. If all `__torch_function__` implementations return `NotImplemented`, a
+   `TypeError` is raised with an appropriate error message.
+
+In practice, for most PyTorch functions, the list of tensor-likes is already
+available and the dispatcher doesn't need to be called. Additionally, while
+equivalent to the code above, if the `Tensor`-likes are all `Tensor` or don't have
+an `__torch_function__` implementation, the internal implementation is called
+immediately. This is done as a performance optimisation to avoid overhead for
+concrete `Tensor` objects.
+
+It will be the job of the dispatcher to extract `Tensor`-like objects from the
+argument list, however, arguments of type `Optional[Tensor]` will be considered
+`Tensor`-like. If one gets a compound or dependent type such as `List[Tensor]`
+or `Tuple[Tensor, ...]` or `Tuple[Tensor, int]`, the dispatcher will have the job
+of extracting an iterable of objects that *could* be `Tensor`-like.
+
+### Generic implementation of `__torch_function__`
+`torch.Tensor` will gain a generic `__torch_function__` of the following form:
+
+```python
+class Tensor:
+    def __torch_function__(self, func, types, args, kwargs):
+        if not all(issubclass(type(self), t) for t in types):
+            return NotImplemented
+        
+        # Defer to internal implementation
+        ret = func._implementation(*args, **kwargs)
+        if type(self) is not Tensor and isinstance(ret, Tensor):
+            ret = Tensor.make_subclass(ret, type(self))
```

Comment:
Does this work for non-methods, e.g., `torch.function`?

Reply:
It does.